upstream/upstream.go

```diff
@@ -654,11 +658,83 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 		}
 	}
 
-	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	// if block can be determined from request and upstream is only full-node and block is historical skip
+	bn, ebn := req.EvmBlockNumber()
+	if ebn != nil || bn <= 0 {
+		return fmt.Errorf("failed to get valid request block number: %w", ebn), false
+	}
+
+	if u.config.Evm.NodeType == common.EvmNodeTypeFull {
+		lb := req.Network().EvmStatePollerOf(u.Config().Id).LatestBlock()
```

Comment:
We probably should check for nil of req.Network() (and maybe evm state poller too)

On initialization when we're detecting chain id of an upstream network is not assigned yet because that's what we're detecting. I guess this situation happens when upstream is defined but evm.chainId is not defined in the config.